### ISSUE TITLE
C1901 (compare-to-empty-string) message is not explicit about the problem and the solution

### ISSUE THREAD

**Issue by kfsone** [PyCQA/pylint#5234]
### Bug description

Consider

```python
   val_type, val_value = get_type(name)  # returns str, Any
   if val_type is Environment:
     if val_value is None or val_value is "":  # or ==
```

The last line produces a pylint error, with the common guidance (e.g https://vald-phoenix.github.io/pylint-errors/plerr/errors/compare-to-empty-string/C1901.html) suggesting that you should use:

```python
    if val_value is None or not len(val_value):
```

The user used `val_value is ""` to explicitly say: is it a string, and is it empty?

To avoid doing this `is` or `==` comparison, the user would need to write:

```python
    if val_value is None or (isinstance(val_value, str) and not len(val_value)):
```

example of following the guidance too literally:

```python
    def fn(v):
      if len(v):
          return "empty"
      return "other"

    fn(None)    # OK: returns "other"
    fn("a")     # OK: returns "other"
    fn([None])  # OK: returns "other"
    fn("")      # OK: returns "empty"
    fn([])      # WRONG: returns "other"
    fn(0)       # ERROR: int has no len
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
a.py.3.1: [C1901(compare-to-empty-string), a.py] Avoid comparisons to empty string [229,1]
```

### Expected behavior

This should either be disabled by default or it should provide better guidance. Why is `x is ""` bad?

### Pylint version

```shell
pylint 1.9.5,
astroid 1.6.6
Python 2.7.18 (v2.7.18:8d21aa21f2, Apr 20 2020, 13:25:05) [MSC v.1500 64 bit (AMD64)]
```

**Comment by Pierre-Sassoulas** [comment]
Hi, thank you for opening the issue. We're not supporting python 2, anymore, does the problem still exists in latest versions ?

**Comment by Pierre-Sassoulas** [comment]
Btw I would use bool, not len personally here.

**Comment by DudeNr33** [comment]
Pylint encourages [Truth Value Testing](https://docs.python.org/3.10/library/stdtypes.html#truth-value-testing), which also applies to strings. That's why ``x is ""`` is considered "bad", as it is redundant - at least as long as you don't want to explicitly check the type. 

Disabling it by default would lead to a lot of redundant and (arguably) "unpythonic" comparisons to go unnoticed, while this here seems like a rather special case.

**Comment by Pierre-Sassoulas** [comment]
The message is now (in 2.11) ``a.py:3:25: R0123: Comparison to literal (literal-comparison)``, which I think could be made more explicit.

**Comment by DudeNr33** [comment]
Maybe we should include the words "Truth Value Testing" as this is the official wording in the Python docs, and people who want to know more can therefore easily find the page linked above.

``'if val_value is ""' could be replaced by 'if val_value' (Truth Value Testing) (literal-comparison)``

**Comment by Pierre-Sassoulas** [comment]
Flake8 says : ``F632 use ==/!= to compare constant literals (str, bytes, int, float, tuple)``. we can also piggy back on the documentation for this flake8 check when pylint users search for the error messages.

**Comment by DudeNr33** [comment]
I think we might be discussing separate things here. 🤔 
The original title and original problem description was about C1901 ``compare-to-empty-string``.

The title is now changed to refer to R0123 ``literal-comparison`` (I also mixed those up in my last response).

I agree that flake8's help message for the R0123 equivalent is better and pylint could adopt it.
But that would not solve the original problem, or am I missing something?
C1901 is in the optional ``CompareToEmptyStringChecker``, and if enabled fires alongside R0123 for this code.

**Comment by Pierre-Sassoulas** [comment]
I ran pylint 2.12.0(dev0 on:

```python
val_type, val_value = get_type(name)  # returns str, Any
if val_type is Environment:
    if val_value is None or val_value is "":  # or ==
        pass  
```
And the closest I got to something unclear was ``R0123``. So I'm assuming the original issue with C1901 does not exists in the latest pylint and we're not supporting pylint 1.9 for python 2.7 anymore.

**Comment by DudeNr33** [comment]
It exists in ``pylint.extensions.emptystring.CompareToEmptyStringChecker`` and is an optional checker. If enabled through ``--load-plugins=pylint.extensions.emptystring`` the message C1901 fires.

**Comment by Pierre-Sassoulas** [comment]
My bad. You're right it's still in current pylint. Now I think this check is redundant with ``use-implicit-booleaness-not-x`` checkers.

**Comment by kfsone** [comment]
I did have extensions.emptystring on at the start of the 2to3 migration I'm doing, so I may have had some conflation of Python's new SyntaxWarning on trying to distinguish "absent value" vs "value absent" with PyLint's message, but it was ultimately C1901 compare-to-empty-string that I turned off to avoid this.

As far as I can divine, all empty `str`s are the same empty `str`, and all integer 0s are the same integer zero, so telling users not to test for 0 or an empty string with `is` seems dangerous guidance. The classic example being None vs "" when looking up a config value. But also, frequently, integer values are "optional":

```
    rc = popen_process.poll()
    if not rc:  # rc == 0
        return "Success"
    ...
```

But poll() returns None to indicate success, not True.

```
    rc = popen_process.poll()
    if rc is 0:
        return "Succees"
    if rc:
        raise Exception(f"process terminated with {rc}")
    raise Exception("process hung")
````

**Comment by Pierre-Sassoulas** [comment]
Well, using truth value is relatively safe as long as you don't mix types together. If a variable can be int, str, a list or a dict then sure, you should use 'is'. But I don't think the code is ideal to understand, or maintain either (ie.you probably have more pressing problems than fixing pylint convention message)

**Comment by clavedeluna** [comment]
Given the back and forth discussion above, what's the current solution to this issue? Updating the message how?

**Comment by Pierre-Sassoulas** [comment]
I think this might be blocked/related to https://github.com/PyCQA/pylint/pull/6870. There's some rationalization to do regarding those messages before making them better.

**Comment by Pierre-Sassoulas** [comment]
Ok, so something generic could be:
```
use-implicit-booleaness-not-comparison: "empty_tuple < ()" can be simplified to "not empty_tuple" as an empty tuple is falsey
use-implicit-booleaness-not-comparison"empty_list <= []" can be simplified to "not empty_list" as an empty list is falsey
use-implicit-booleaness-not-comparison:"Y is not ''" can be simplified to "Y" as an empty string is falsey
use-implicit-booleaness-not-comparison-to-zero: "Y is not 0" can be simplified to "Y" as 0 is falsey
```
